### PR TITLE
Allows to automate authentication via environment variables.

### PR DESCRIPTION
**Pull request by jfboismenu**
tk-run-app can now do script and user based authentication from the command line. We've also exposed the authentication method to the outside world so tests can leverage it.

### CODE REVIEW

File: tk_toolchain/authentication.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import os
+
+
+def _get_toolkit_user(sg_auth, environment):
+    """
+    Retrieves the Toolkit user using the passed in authenticator and
+    environment.
+    """
+
+    host = environment.get("TK_TOOLCHAIN_HOST")
```
Comment (jfboismenu) [inline tk_toolchain/authentication.py:21]: moved from `tk_run_app`

File: tests/test_authentication.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        args[method] = Mock(spec_list=[])
+        args[method].side_effect = lambda *args, **kwargs: set(
+            list(args) + list(kwargs.values())
+        )
+    return Mock(spec_list=list(args.keys()), **args)
+
+
+def test_user_based_auth(sg_auth_mock):
+    """
+    Test used based auth.
```
Comment (pscadding) [inline tests/test_authentication.py:34]:
```suggestion
    Test user based auth.
```

File: tests/test_authentication.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        "TK_TOOLCHAIN_USER_PASSWORD": "hailToTheKing",
+    }
+    result = _get_toolkit_user(sg_auth_mock, env)
+    assert sg_auth_mock.create_session_user.called
+    assert result == set(env.values())
+
+
+def test_user_overrides_script(sg_auth_mock):
+    """
+    Test used based auth.
```
Comment (pscadding) [inline tests/test_authentication.py:48]:
```suggestion
    Test user based auth.
```

File: tests/test_authentication.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    all_env = {
+        "TK_TOOLCHAIN_SCRIPT_NAME": "automation",
+        "TK_TOOLCHAIN_SCRIPT_KEY": "!$3,asdas$dfadf*(0",
+    }
+    all_env.update(user_env)
+    result = _get_toolkit_user(sg_auth_mock, all_env)
+
+    assert sg_auth_mock.create_session_user.called
+    assert result == set(user_env.values())
```
Comment (pscadding) [inline tests/test_authentication.py:63]: I'm not quite sure I understand the purpose of this test. Is it to test that when both script and user credentials are provided, the user credentials take precedence?

Comment (jfboismenu) [reply tests/test_authentication.py:63]: Yes. I've updated the doc string to make the intent clearer.

**Pull request merged**